sql/2021/privacy/most_common_cmps_for_iab_tcf_v2.sql

```diff
@@ -0,0 +1,42 @@
+#standardSQL
+# Counts of CMPs using IAB Transparency & Consent Framework
+# cf. https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#tcdata
+
+WITH pages_iab_tcf_v2 AS (
+  SELECT
+    _TABLE_SUFFIX AS client,
+    JSON_QUERY(payload, "$._privacy.iab_tcf_v2.data") AS metrics
+  FROM
+    `httparchive.pages.2021_07_01_*`
+  WHERE
+    JSON_QUERY(metrics, "$.iab_tcf_v2.data") IS NOT NULL
+),
+
+total_nb_pages AS (
+  SELECT
+    _TABLE_SUFFIX AS client,
+    COUNT(0) AS total_nb_pages
+  FROM
+    `httparchive.pages.2021_07_01_*`
+  GROUP BY
+    client
+)
+
+SELECT
+  client,
+  JSON_VALUE(metrics, '$.cmpId') AS cmpId,
+  COUNT(0) AS nb_websites,
+  COUNT(0) / MIN(total_nb_pages.total_nb_pages) AS pct_websites
+FROM
+  pages_iab_tcf_v2
+JOIN
+  total_nb_pages
+USING (client)
+WHERE
+  JSON_VALUE(metrics, '$.cmpId') IS NOT NULL
+GROUP BY
+  client,
+  cmpId
+ORDER BY
+  client ASC,
+  nb_websites DESC
```

Comment:
This query no longer runs :-(

The problem is the `_privacy` is not real JSON but escaped JSON. So you need to use this:

```sql
    JSON_QUERY(JSON_VALUE(payload, "$._privacy"), "$.iab_tcf_v2.data") AS metrics
```

as opposed to

```sql
   JSON_QUERY(payload, "$._privacy.iab_tcf_v2.data") AS metrics
```

Also your first `WHERE` clause is incorrect.

I also like to include the total in the final query (it's a nice spot check for `OVER` issue).

I actually think this query can be simplified further, so including the above changes, we can simply do something like this with only one CTE:

```suggestion
WITH total_pages AS (
  SELECT
    _TABLE_SUFFIX,
    COUNT(0) AS total
  FROM
    `httparchive.pages.2021_07_01_*`
  GROUP BY
    _TABLE_SUFFIX
)

SELECT
  _TABLE_SUFFIX AS client,
  JSON_QUERY(JSON_VALUE(payload, "$._privacy"), "$.iab_tcf_v2.data.cmpId") AS cmpId,
  COUNT(0) AS number_websites,
  total,
  COUNT(0) / MIN(total) AS pct_websites
FROM
  `httparchive.pages.2021_07_01_*`
JOIN
  total_pages
USING (_TABLE_SUFFIX)
WHERE
  JSON_QUERY(JSON_VALUE(payload, "$._privacy"), "$.iab_tcf_v2.data.cmpId") IS NOT NULL
GROUP BY
  client,
  total,
  cmpId
ORDER BY
  client ASC,
  number_websites DESC
```

WDYT? Can you check the other queries for similar issues and, make similar changes to them? Happy to use multiple CTEs where it simplifies the query, but don't think it does in this case to be honest. Some of your other queries are more complicated so may make more sense for them.

Reply:
Yeah, looks like I left most queries using custom metrics in a horribly broken state :sweat_smile: 
For simple queries, the simplification is fine. However, for a query like https://github.com/HTTPArchive/almanac.httparchive.org/blob/c642950b56caddc3894e5738a841c2559188ca50/sql/2021/privacy/nb_websites_with_features_based_on_string_search.sql, I wonder whether repeatedly extracting `JSON_VALUE(payload, "$._privacy")` (everywhere where it incorrectly says `metrics` now) instead of just once in a CTE could reduce performance?

Reply:
Agreed. Use CTEs where appropriate but don’t unnecessarily use them for simple queries.

And btw all our custom metrics used to be written this way so not your fault. It’s only more recently (e.g. PWA custom metric) that we’ve changed how to do it in a way that saves them as proper JSON instead of an escaped string.

Reply:
Done in b3cedee and ef44cab.